File: src/CircleAI.Memory/MemoryAtom.cs
// MemoryAtom.cs
//
// One remembered thing, of one kind, traceable to where it was said.
//
// THE LAYER THE STORE DID NOT HAVE. Episodes (L0) hold what was said and a
// persona (L3) holds a summary of who somebody is, with nothing in between —
// so anything specific enough to act on had to be re-derived from raw turns
// every time, which in practice means it was not recalled at all.
//
// AN ATOM IS NOT A RULE. Nothing here blocks anything. The kind decides how an
// atom is WEIGHTED when several match the moment, and whether it decays; a
// fact carries how to re-check itself, and a stale fact is reported rather
// than obeyed. Recall, not handcuffs.
//
// NEVER DELETED, ONLY SUPERSEDED. The correction history is the signal that
// makes one atom outrank another - "you have said this four times" is a
// different fact from "you said this once" - and overwriting throws away
// exactly the thing that makes a memory feel load-bearing.

using System;

namespace CircleAI.Memory;

/// <summary>What kind of thing is being remembered.</summary>
/// <remarks>
/// The four behave differently enough that one undifferentiated pile
/// guarantees mishandling: a preference treated as a ruling makes the assistant
/// rigid, and a ruling treated as a preference makes it useless.
/// </remarks>
public enum AtomKind
{
    /// <summary>
    /// Something that came up, what was chosen, and how it turned out.
    /// </summary>
    /// <remarks>
    /// THE FIRST KIND WORTH HAVING, and the one that needs no judgement to
    /// write down. Every other kind asks a classification question at the
    /// moment of capture - is this a ruling or a preference? - and the answer
    /// is exactly the thing most likely to be got wrong by whoever is closest
    /// to the mistake. A decision is a record: what happened, what was decided,
    /// whether it worked.
    /// <para>
    /// The failures are worth as much as the fixes. "Tried adb push, it wrote
    /// nothing" saves the next attempt as surely as knowing what did work.
    /// </para>
    /// </remarks>
    Decision,

    /// <summary>A decision that was made. Never decays; surfaces first.</summary>
    Ruling,

    /// <summary>Something true about the world. Re-checked before it is relied on.</summary>
    Fact,

    /// <summary>How somebody likes things done. Applied by default, easy to override.</summary>
    Preference,

    /// <summary>
    /// How to work with this person. Never quoted back at them - it shapes tone
    /// and how much to ask, which is not the same as being repeated.
    /// </summary>
    Relationship,
}

/// <summary>One fact, one kind, one source.</summary>
public sealed class MemoryAtom
{
    /// <summary>Stable identity, so a later atom can point at the one it replaces.</summary>
    public Guid Id { get; init; } = Guid.NewGuid();

    /// <summary>What sort of thing this is.</summary>
    /// <remarks>
    /// Deliberately an open dimension. Decision is the kind being used now;
    /// the rest are room to grow into, and a kind added later costs a value
    /// here and a weight in the ranking rather than a second store.
    /// </remarks>
    public AtomKind Kind { get; init; } = AtomKind.Decision;

    /// <summary>The thing itself, in the words a person would recognise.</summary>
    public string Text { get; init; } = string.Empty;

    /// <summary>
    /// What this is about, as a situation key - "deploy:android", "language:count".
    /// </summary>
    /// <remarks>
    /// THIS IS WHAT MAKES RECALL-AT-THE-MOMENT POSSIBLE. Searching prose for
    /// relevance is a guess; matching the subject of the action against the
    /// subject of the atom is not. Free text still gets searched, but the
    /// subject is what makes the right atom arrive before the wrong one.
    /// </remarks>
    public string? Subject { get; init; }

    /// <summary>The episode this came out of, so the claim can be walked back.</summary>
    /// <remarks>
    /// A memory that cannot be audited is a rumour. Every layer above the raw
    /// turns keeps a path down to them.
    /// </remarks>
    public Guid? SourceEpisode { get; init; }

    /// <summary>When it was first recorded.</summary>
    public DateTimeOffset RecordedAtUtc { get; init; } = DateTimeOffset.UtcNow;

    /// <summary>
    /// How many times this had to be corrected into place.
    /// </summary>
    /// <remarks>
    /// THE COUNT IS THE POINT, and it is a relevance signal rather than a
    /// punishment ledger: something corrected three times in a week is what
    /// somebody most needs put in front of them, whatever else also matches.
    /// </remarks>
    public int Corrections { get; init; }

    /// <summary>When it was last corrected, or null if it never was.</summary>
    public DateTimeOffset? LastCorrectedUtc { get; init; }

    /// <summary>The atom that replaced this one, or null while it still stands.</summary>
    public Guid? SupersededBy { get; init; }

    /// <summary>Whether this atom is still the current answer.</summary>
    public bool IsCurrent => SupersededBy is null;

    /// <summary>
    /// For a decision: what came up that had to be decided.
    /// </summary>
    /// <remarks>
    /// Kept apart from <see cref="Text"/> because the two answer different
    /// questions - "have we been here before" is matched against this, and
    /// "what did we do about it" is read out of that. Null for kinds that are
    /// not a decision.
    /// </remarks>
    public string? Challenge { get; init; }

    /// <summary>For a decision: how it turned out.</summary>
    public DecisionOutcome? Outcome { get; init; }

    /// <summary>
    /// For a fact: how to re-check it against reality.
    /// </summary>
    /// <remarks>
    /// A command, a path, a query - whatever the caller can act on. This is the
    /// ONE piece of executable memory in the design and it is deliberately a
    /// verification rather than an assertion: it decides whether a fact is
    /// trustworthy, never whether an action is allowed.
    /// </remarks>
    public string? Verify { get; init; }

    /// <summary>When the fact was last checked.</summary>
    public DateTimeOffset? VerifiedAtUtc { get; init; }

    /// <summary>What the last check found, or null if it has never run.</summary>
    public bool? VerifiedOk { get; init; }

    /// <summary>
    /// Whether this should be shown with a warning rather than relied on.
    /// </summary>
    /// <remarks>
    /// A fact that failed its check still surfaces - WITH the fact that it
    /// failed. Hiding it would leave somebody acting on the stale belief they
    /// already had, which is worse than showing a doubt.
    /// </remarks>
    public bool IsStale => Kind == AtomKind.Fact && VerifiedOk == false;

    /// <summary>A decision that was tried and did not work.</summary>
    /// <remarks>
    /// Surfaced as loudly as one that did: the point of remembering a failure
    /// is to stop the next attempt taking the same road.
    /// </remarks>
    public bool Failed => Outcome == DecisionOutcome.Failed;
}

/// <summary>How a decision turned out.</summary>
public enum DecisionOutcome
{
    /// <summary>Decided, but nobody has found out yet whether it worked.</summary>
    Open,

    /// <summary>It worked. This is the road to take again.</summary>
    Resolved,

    /// <summary>It did not. Worth as much as a fix, and often sooner.</summary>
    Failed,
}

File: kotlin/src/main/kotlin/com/bhengubv/circleai/memory/MemoryAtom.kt
// MemoryAtom.kt
//
// The atom: one thing worth remembering, and the situation that finds it.
//
// Port of MemoryAtom.cs, Situation.cs, IAtomStore.cs, AtomCandidate.cs and
// IAtomExtractor.cs.

package com.bhengubv.circleai.memory

import java.time.Instant
import java.util.UUID

enum class AtomKind { DECISION, RULING, FACT, PREFERENCE, RELATIONSHIP }

enum class DecisionOutcome { OPEN, RESOLVED, FAILED }

data class MemoryAtom(
    val id: UUID = UUID.randomUUID(),
    val kind: AtomKind = AtomKind.DECISION,
    val text: String = "",
    val subject: String? = null,
    val sourceEpisode: UUID? = null,
    val recordedAtUtc: Instant = Instant.EPOCH,
    val machine: String? = null,
    val corrections: Int = 0,
    val lastCorrectedUtc: Instant? = null,
    val supersededBy: UUID? = null,
    val challenge: String? = null,
    val outcome: DecisionOutcome? = null,
    val verify: String? = null,
    val verifiedAtUtc: Instant? = null,
    val verifiedOk: Boolean? = null,
) {
    val isCurrent: Boolean get() = supersededBy == null

    /** A FACT that failed its own check. Still readable, no longer an answer. */
    val isStale: Boolean get() = kind == AtomKind.FACT && verifiedOk == false

    val failed: Boolean get() = outcome == DecisionOutcome.FAILED
}

/**
 * What is about to happen, described well enough to look it up.
 *
 * THIS IS THE WHOLE DIFFERENCE between a memory that helps and one that does
 * not. Loading everything at the start of a conversation puts the rules
 * furthest from the moment they apply; an hour and forty tool calls later,
 * nothing read at the greeting is meaningfully present, and no amount of
 * emphasis in the file changes that.
 *
 * So recall is keyed on the ACTION rather than the session. Before a deploy,
 * ask what is known about deploying. The subject of the action is matched
 * against the subject of the atom, which is a lookup rather than a guess.
 */
data class Situation(
    val verb: String? = null,
    val target: String? = null,
    val tool: String? = null,
    val text: String? = null,
) {
    val key: String
        get() = listOfNotNull(verb, target)
            .filter { it.isNotBlank() }
            .joinToString(":") { it.trim().lowercase() }

    /**
     * Most specific first, then broader.
     *
     * A slash-delimited target is walked UP - android/p30 also matches android -
     * because a rule filed against the general case has to be found by the
     * specific one. Without that, a rule about deploying to Android is invisible
     * the moment somebody names the phone.
     */
    val keys: List<String>
        get() {
            val out = mutableListOf<String>()
            val v = verb?.trim()?.lowercase()
            var t = target?.trim()?.lowercase()

            if (!v.isNullOrEmpty() && !t.isNullOrEmpty()) {
                out.add(v + ":" + t)
                var cut = t.lastIndexOf('/')
                while (cut > 0) {
                    t = t!!.substring(0, cut)
                    out.add(v + ":" + t)
                    cut = t!!.lastIndexOf('/')
                }
            }

            if (!v.isNullOrEmpty()) out.add(v)
            return out
        }

    val query: String
        get() = listOfNotNull(verb, target, tool, text)
            .filter { it.isNotBlank() }
            .joinToString(" ") { it.trim() }

    val isEmpty: Boolean get() = keys.isEmpty() && text.isNullOrBlank()
}

data class RecallResult(
    val atoms: List<MemoryAtom>,
    val tone: List<MemoryAtom>,
    val considered: Int,
) {
    val any: Boolean get() = atoms.isNotEmpty()

    companion object { val empty = RecallResult(emptyList(), emptyList(), 0) }
}

/**
 * How much recall is allowed to say.
 *
 * A budget rather than a limit: what fits is chosen by rank, so the cap costs
 * the least useful atoms rather than truncating the most useful one mid-word.
 */
data class RecallBudget(val maxAtoms: Int = 5, val maxCharacters: Int = 600) {
    companion object { val default = RecallBudget() }
}

/**
 * Reading and writing the layer between raw turns and a persona.
 *
 * NOTHING HERE REQUIRES AN EMBEDDING. Vector search improves recall; it must
 * never be what ENABLES it. A store that stops working without a 100 MB model
 * is a store that does not work on the phone this is for.
 */
interface IAtomStore {
    suspend fun add(atom: MemoryAtom)
    suspend fun supersede(oldAtomId: UUID, replacement: MemoryAtom): MemoryAtom
    suspend fun match(situation: Situation, limit: Int = 20): List<MemoryAtom>
    suspend fun byKind(kind: AtomKind, limit: Int = 50): List<MemoryAtom>
    suspend fun all(includeSuperseded: Boolean = false, limit: Int = 500): List<MemoryAtom>
    suspend fun knows(text: String): Boolean
    suspend fun get(id: UUID): MemoryAtom?
    suspend fun markVerified(id: UUID, ok: Boolean, whenUtc: Instant)
    suspend fun count(): Int
}

/**
 * Something worth remembering, SPOTTED rather than written.
 *
 * Extraction PROPOSES; it does not decide. A candidate carries what was
 * spotted, which words triggered it and how sure that is, because an extractor
 * that silently writes whatever it thinks it saw fills the memory with noise -
 * and noise ranks. Recall then puts a misreading in front of somebody at the
 * exact moment they are about to act on it, which is worse than an empty
 * memory.
 */
data class AtomCandidate(
    val atom: MemoryAtom,
    val confidence: Double,
    val cue: String,
    val quote: String,
) {
    val certain: Boolean get() = confidence >= RECORD_ABOVE

    companion object {
        /** Above this it is recorded; below it, it is offered. Nothing is superseded on a guess. */
        const val RECORD_ABOVE = 0.80
    }
}

/**
 * The seam between what was said and what is remembered.
 *
 * ONE SEAM, TWO MECHANISMS. CueExtractor needs no model and therefore works on
 * a phone with the radios off, which makes it the FLOOR rather than the
 * fallback. A model reads a conversation better than any list of phrases will,
 * and when one is loaded it should do this job - but it must never be what
 * makes the memory work at all.
 */
interface IAtomExtractor {
    val name: String
    fun extract(episode: EpisodicMemoryEntry, subject: String? = null): List<AtomCandidate>
}
